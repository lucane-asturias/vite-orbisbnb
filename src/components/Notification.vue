<template>
  <div v-if="notification.active"
    class="notification flex justify-between p-2 mb-7"
    :class="{ 'bg-blackish': darkMode, 'bg-whitesmoke': !darkMode }"  
  >

    <p 
      class="ml-3 text-sm tracking-wide" 
      :class="{ 'text-teal-400': darkMode, 'text-blue-600': !darkMode }"  
      v-text="notification.message"
    />

    <button class="x" @click="toggleNotification" />

  </div>
</template>

<script lang="ts" setup>
  import useDarkMode from '../composables/useDarkMode';

  type Notification = { message: string; active: boolean }

  interface Props {
    notification: Notification
    toggleNotification: Function
  }

  defineProps<Props>()

  const { darkMode } = useDarkMode()
</script>

<style>

.notification { animation: fadeIn 1.5s; }

@keyframes fadeIn {
  0% { opacity: 0; }
  100% { opacity: 1; }
}

/* CSS to draw a X */

.x {
  background-color: rgba(10, 10, 10, 0.3);
  border-radius: 50%;
  cursor: pointer;
  pointer-events: auto;
  height: 20px;
  width: 20px;
  outline: none;
  position: relative;
}

.x::before, .x::after {
  background-color: white;
  content: "";
  display: block;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translateX(-50%) translateY(-50%) rotate(45deg);
  transform-origin: center center;
}

.x::before {
  height: 2px;
  width: 50%;
}

.x::after {
  height: 50%;
  width: 2px;
}

.x:hover {
  background-color: rgba(10, 10, 10, 0.6);
}

</style>