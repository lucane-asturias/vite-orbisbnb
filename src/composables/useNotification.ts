import { reactive } from 'vue'

const data = reactive({
  message: '',
  active: false
})

const useNotification = () => {
  
  const setNotification = (newMessage: string) => {
    data.message = newMessage
    data.active = true
    setTimeout(() => data.active = false, 30000)
  }

  const toggleNotification = () => {
    data.active = !data.active
    setTimeout(() => data.active = !data.active, 30000)
  }

  return {
    notification: data,
    setNotification,
    toggleNotification
  }
}

export default useNotification
