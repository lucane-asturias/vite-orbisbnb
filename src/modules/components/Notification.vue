<template>
  <div v-if="notification.active" 
    class="flex justify-between bg-whitesmoke p-2 mb-4"
    :class="{ 'is-primary': darkMode, 'is-info': !darkMode }">

    <p class="ml-3 text-sm text-blue-600 tracking-wide" v-text="notification.message" />

    <button class="teste" @click="toggleNotification" />

  </div>
</template>

<script lang="ts" setup>
  import useDarkMode from '../hooks/useDarkMode';

  type Notification = { message: string; active: boolean }

  interface Props {
    notification: Notification
    toggleNotification: Function
  }

  defineProps<Props>()

  const { darkMode } = useDarkMode()
</script>

<style>

/* CSS to draw a X */

.teste {
  background-color: rgba(10, 10, 10, 0.3);
  border-radius: 50%;
  cursor: pointer;
  pointer-events: auto;
  height: 20px;
  width: 20px;
  outline: none;
  position: relative;
}

.teste::before, .teste::after {
  background-color: white;
  content: "";
  display: block;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translateX(-50%) translateY(-50%) rotate(45deg);
  transform-origin: center center;
}

.teste::before {
  height: 2px;
  width: 50%;
}

.teste::after {
  height: 50%;
  width: 2px;
}

.teste:hover {
  background-color: rgba(10, 10, 10, 0.6);
}

</style>