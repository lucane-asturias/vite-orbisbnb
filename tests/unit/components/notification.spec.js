import { shallowMount } from '@vue/test-utils'
import Notification from '@/components/Notification.vue'
import useNotification from '@/composables/useNotification'

describe('Tests on the Notification.vue component', () => {
  let wrapper
  // SETUP - run prior to each unit test
  beforeEach(() => {
    vi.useFakeTimers() // tell vitest we use mocked time
    // render the component
    wrapper = shallowMount(Notification, {
      props: {
        notification: { message: 'Testing Notification', active: true },
        toggleNotification: vi.fn()
      }
    })
  })

  test('should render the component correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should dynamically render Notification.vue', async () => {
    expect(wrapper.find('.notification').exists()).toBeTruthy()
    expect(wrapper.find('p').text()).toMatch('Testing Notification')

    await wrapper.setProps({ notification: { active: false } })

    expect(wrapper.find('.notification').exists()).toBeFalsy()
    expect(wrapper.find('p').exists()).toBeFalsy()
  })

  test('notification composable should work correctly', () => {
    const { notification, setNotification, toggleNotification } = useNotification()

    // default values
    expect(typeof notification.message).toBe('string')
    expect(notification.message).toBe('')
    expect(notification.active).toBeTypeOf('boolean')
    expect(notification.active).toBeFalsy()

    // functions
    setNotification('Testing Notification')
    expect(notification.message).toMatch('Testing Notification')
    expect(notification.active).toBeTruthy()

    toggleNotification()
    expect(notification.active).toBeFalsy()
    vi.advanceTimersByTime(30000)
    expect(notification.active).toBeTruthy()
  })

})