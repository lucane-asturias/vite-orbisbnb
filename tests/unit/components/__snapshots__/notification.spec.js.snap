// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on the Notification.vue component > should render the component correctly 1`] = `
"<div class=\\"notification flex justify-between p-2 mb-7 bg-whitesmoke\\">
  <p class=\\"ml-3 text-sm tracking-wide text-blue-600\\">Testing Notification</p><button class=\\"x\\"></button>
</div>"
`;
